src/Libraries/Microsoft.Extensions.AI/ToolReduction/EmbeddingToolReductionStrategy.cs

```diff
@@ -0,0 +1,212 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Numerics.Tensors;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+#pragma warning disable IDE0032 // Use auto property, suppressed until repo updates to C# 14
+
+/// <summary>
+/// A tool reduction strategy that ranks tools by embedding similarity to the current conversation context.
+/// </summary>
+/// <remarks>
+/// The strategy embeds each tool (name + description by default) once (cached) and embeds the current
+/// conversation content each request. It then selects the top <c>toolLimit</c> tools by similarity.
+/// </remarks>
+[Experimental("MEAI001")]
+public sealed class EmbeddingToolReductionStrategy : IToolReductionStrategy
+{
+    private readonly ConditionalWeakTable<AITool, Embedding<float>> _toolEmbeddingsCache = new();
+    private readonly IEmbeddingGenerator<string, Embedding<float>> _embeddingGenerator;
+    private readonly int _toolLimit;
+
+    private Func<AITool, string> _toolEmbeddingTextFactory = static t =>
+    {
+        if (string.IsNullOrWhiteSpace(t.Name))
+        {
+            return t.Description;
+        }
+
+        if (string.IsNullOrWhiteSpace(t.Description))
+        {
+            return t.Name;
+        }
+
+        return t.Name + "\n" + t.Description;
+    };
+
+    private Func<IEnumerable<ChatMessage>, string> _messagesEmbeddingTextFactory = static messages =>
+    {
+        var messageTexts = messages.Select(m => m.Text).Where(s => !string.IsNullOrEmpty(s));
+        return string.Join("\n", messageTexts);
+    };
+
+    private Func<ReadOnlyMemory<float>, ReadOnlyMemory<float>, float> _similarity = static (a, b) => TensorPrimitives.CosineSimilarity(a.Span, b.Span);
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EmbeddingToolReductionStrategy"/> class.
+    /// </summary>
+    /// <param name="embeddingGenerator">Embedding generator used to produce embeddings.</param>
+    /// <param name="toolLimit">Maximum number of tools to return. Must be greater than zero.</param>
+    public EmbeddingToolReductionStrategy(
+        IEmbeddingGenerator<string, Embedding<float>> embeddingGenerator,
+        int toolLimit)
+    {
+        _embeddingGenerator = Throw.IfNull(embeddingGenerator);
+        _toolLimit = Throw.IfLessThanOrEqual(toolLimit, min: 0);
+    }
+
+    /// <summary>
+    /// Gets or sets a delegate used to produce the text to embed for a tool.
+    /// Defaults to: <c>Name + "\n" + Description</c> (omitting empty parts).
+    /// </summary>
+    public Func<AITool, string> ToolEmbeddingTextFactory
```

Comment:
It's not wrong, but "Factory" feels odd to me personally as the name for these properties.